oop-3-inheritance_nd_visibility/index.md

```diff
@@ -0,0 +1,454 @@
+
+In Java, inheritance is an is-a relationship. That is, we use inheritance only if there exists an is-a relationship between two classes. For example,
+
+Laptop is a Prouct Car is a Vehicle Orange is a Fruit
+
+## Why do we need it?
+
+- The most important use of inheritance in Java is code reusability. The code that is present in the parent class can be directly used by the child
+  class.
+- Method overriding is also known as runtime polymorphism. Hence, we can achieve Polymorphism in Java with the help of inheritance.
+
+## Have we seen it before?
+
+**YES!!** Since day one actually. Remember this example from the very first lesson?
+
+```java
+package com.redi.j2;
+
+public class Main {
+
+    public static void main(String[] args) {
+        System.out.println("Hello World");
+
+        String name = "Memet";
+        String profession = "skydiver";
+        int age = 40;
+        boolean smoking = false;
+
+        coolOrNotCool(profession);
+
+        for (int i = 0; i < 10; i++) {
+            System.out.println("hey " + i);
+        }
+    }
+
+
+    private static void coolOrNotCool(String profession) {
+        if (profession.equals("skydiver")) {
+            System.out.println("Cool");
+        }
+        else {
+            System.out.println("Not cool");
+        }
+    }
+}
+```
+
+let's take a moment and understand what is a [`String`](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html) really?
+
+## Visibility and access modifiers
+
+### Default Access Modifier - No Keyword
+
+- Default access modifier means we do not explicitly declare an access modifier for a class, field, method, etc.
+
+- A variable or method declared without any access control modifier is available to any other class in the same package. The fields in an interface
+  are implicitly public static final and the methods in an interface are by default public.
+
+### Private Access Modifier - Private
+
+- Methods, variables, and constructors that are declared private can only be accessed within the declared class itself.
+
+- Private access modifier is the most restrictive access level. Class and interfaces cannot be private.
+
+### Public Access Modifier - Public
+
+- A class, method, constructor, interface, etc. declared public can be accessed from any other class. Therefore, fields, methods, blocks declared
+  inside a public class can be accessed from any class belonging to the Java Universe.
+
+![modifiers summary](Screenshot%202021-10-13%20113506.png)
+
+## What is _Method Overriding_?
+
+Let's have a look at another example of handy Inheritance: defining geometric shapes.
+
+```java
+public class Shape {
+
+    private final String type;
+
+    private final double area;
+
+    private final double perimeter;
+
+
+    public Shape(final String type, final double area, final double perimeter) {
+        this.type = type;
+        this.area = area;
+        this.perimeter = perimeter;
+    }
+
```

Comment:
I would add a Java comment somewhere (after the constructor?) indicating that all properties have getters.

```java
// Getters for all properties
```